lms/djangoapps/course_api/api.py

```diff
@@ -266,3 +268,103 @@ def get_course_run_url(request, course_id):
+    and the default value is 1000). More than that and the method will raise a
+    `OverEnrollmentLimitException` exception.
+
+    This method works by querying the `CourseEnrollment` and `CourseAccessRole` models,
+    prefetching user information and *then joining results in Python using dictionaries*.
+    This approach was choosen to avoid database heavy queries (such as DISTINCT and COUNT) in
+    the `CourseEnrollment` table, which would take too long to complete in a request lifecycle.
+
+    The main concern with this approach is the dataset size and resource usage since this method
+    returns all enrollments without pagination. We're using a conservative number on the
+    `COURSE_MEMBER_API_ENROLLMENT_LIMIT` setting to avoid any issues.
+
+    Examples:
+        - Get all course members:
+            get_course_members(course_key)
+
+    Arguments:
+        course_key (CourseKey): CourseKey to retrieve student data.
+
+    Returns:
+        dict: A dictionary with the following format:
+            {
+                "user_id": {
+                    "id": 12,
+                    "username": "jonh5000",
+                    "email": "jonh@example.com",
+                    "name": "Jonh Doe",
+                    "enrollment_mode": "verified",
+                    "roles": [
+                        "student",
+                        "instructor",
+                        "staff",
+                    ]
+                }
+            }
+    """
+    def get_user_info(user, enrollment_mode=None):
+        """
+        Utility function to extract user information from model.
+        """
+        return {
+            "id": user.id,
+            "username": user.username,
+            "email": user.email,
+            "name": user.profile.name,
+            "enrollment_mode": enrollment_mode,
+        }
+
+    # Raise error if trying to retrieve user list from a course with more than
+    # settings.COURSE_MEMBER_API_ENROLLMENT_LIMIT active enrollments. The fastest way
+    # to do this is to query for the 1st item after `COURSE_MEMBER_API_ENROLLMENT_LIMIT`.
+    over_limit = CourseEnrollment.get_active_enrollments_in_course(
+        course_key
+    )[settings.COURSE_MEMBER_API_ENROLLMENT_LIMIT:][:1]
+    if over_limit.exists():
+        raise OverEnrollmentLimitException(
+            f"Can't retrieve course members for courses with more than "
+            f"{settings.COURSE_MEMBER_API_ENROLLMENT_LIMIT} active enrollments. "
+            f"This value is retrieved from `settings.COURSE_MEMBER_API_ENROLLMENT_LIMIT`."
+        )
+
+    # Python dicts where we're going to manually combine the data from the two querysets
+    user_roles = defaultdict(list)
+    user_info = {}
+
+    # Retrieve all active enrollments in course and prefetch user information
+    enrollments = CourseEnrollment.get_active_enrollments_in_course(
+        course_key
+    )[:settings.COURSE_MEMBER_API_ENROLLMENT_LIMIT]
```

Comment:
Why are we still limiting here?  If the limit accomplishes anything, then we'd never reach this line of code, right?

Reply:
Yup, makes sense. I copied this from Dave's pseudo-implementation above but it has no effect (or at least it shouldn't).